### ISSUE TITLE
git.latest: rev with tag fails

### ISSUE THREAD

**Issue by jkuettner** [saltstack/salt#23587]
the [doc](http://docs.saltstack.com/en/latest/ref/states/all/salt.states.git.html#salt.states.git.latest) says:

> rev
> The remote branch, tag, or revision ID to checkout after clone / before update

but if I try to put a git-tag into the rev option the git-state tries to run the following command (replace the <TAG> value with a git-tag):
```
git reset --hard origin/<TAG>' failed. Stderr: "fatal: ambiguous argument 'origin/6.2.12': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions
```
the correct command has to be `git reset --hard tags/<TAG>`

So a workaround is to add "tags" to the remote_name option of the git-state but I'm not sure if this is the desired way to go ..

I'm working with the following salt-version:

```
                  Salt: 2015.5.0
                Python: 2.7.3 (default, Mar 13 2014, 11:03:55)
                Jinja2: 2.6
              M2Crypto: 0.21.1
        msgpack-python: 0.1.10
          msgpack-pure: Not Installed
              pycrypto: 2.6
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 13.1.0
                  RAET: Not Installed
                   ZMQ: 3.2.3
                  Mako: Not Installed
 Debian source package: 2015.5.0+ds-1~bpo70+1
```

**Comment by jfindlay** [comment]
@jkuettner, thanks for the report.  Git tags should be handled by the `rev` argument to git.latest.  There is a small probability this is already fixed on develop, but it does seem that you've found a legitimate bug here.

**Comment by multani** [comment]
#24409 is a duplicate of this ticket. Although this one is older than #24409, the newer one has more information so we should probably close #23587?

**Comment by terminalmage** [comment]
git.latest was rewritten in #26341, which fixes this issue. The fix will be in version 2015.8.0.

As of this rewrite, when a SHA1 or tag is specified as the rev, a branch is checked out at that SHA1. You can also now specify the local branch name to use, where before it would just check out a branch with the same name as the rev, which would create ambiguous refs if a tag was checked out.

The state is also smart enough now not to try to make a non-fast-forward change unless ``force_reset`` is set to ``True``, in which case it performs a ``git reset --hard`` to that SHA1 or tag.

Marking this issue as closed.
